cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,417 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package supervisor
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/fulcio"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	"github.com/sigstore/cosign/v2/pkg/oci/static"
+	"github.com/sigstore/rekor/pkg/client"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/cmd/opampsupervisor/supervisor/config"
+)
+
+var (
+	// agentPackageKey is the key used for the top-level package (the agent).
+	// According to the spec, an empty key may be used if there is only one top-level package.
+	// https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#packages
+	agentPackageKey = ""
+
+	lastPackageStatusFileName = "last-reported-package-statuses.proto"
+)
+
+// maxAgentBytes is the max size of an agent package that will be accepted.
+// It is 1 gibibyte.
+const maxAgentBytes = 1024 * 1024 * 1024
+
+// packageManager manages the persistent state of downloadable packages.
+// It persists the packages to the file system.
+// Currently, it only allows for a single top-level package containing the agent
+// to be received.
+type packageManager struct {
+	persistentState *persistentState
+	topLevelHash    []byte
+	topLevelVersion string
+
+	storageDir string
+	agentPath  string
+	checkOpts  *cosign.CheckOpts
+	am         agentManager
+}
+
+type agentManager interface {
+	stopAgentProcess(ctx context.Context) (chan struct{}, error)
+}
+
+func newPackageManager(
+	agentPath,
+	storageDir,
+	agentVersion string,
+	persistentState *persistentState,
+	signatureOpts config.AgentSignature,
+	am agentManager,
+) (*packageManager, error) {
+	// Read actual hash of the on-disk agent
+	f, err := os.Open(agentPath)
+	if err != nil {
+		return nil, fmt.Errorf("open agent: %w", err)
+	}
+
+	h := sha256.New()
+	if _, err = io.Copy(h, f); err != nil {
+		return nil, fmt.Errorf("compute agent hash: %w", err)
+	}
+	agentHash := h.Sum(nil)
+
+	checkOpts, err := createCosignCheckOpts(signatureOpts)
+	if err != nil {
+		return nil, fmt.Errorf("create signature verification options: %w", err)
+	}
+
+	return &packageManager{
+		persistentState: persistentState,
+		topLevelHash:    agentHash,
```

Comment:
It is a bit confusing that we have both `AllPackagesHash` and `topLevelHash`. Are they the same or different? If different it would be nice to explain in a comment somewhere.